Added pyversion option for doctest (issue 3303)

marco-buttu opened the pull request:
This pull request is related to the   #3303 . Now it is possible to specify the Python version required to execute the doctest. For instance:

```
.. doctest::
   :pyversion: >= 3.3
```

The operator could be: <, <=, ==, >=, >. I also added two tests.

marco-buttu committed:
Added pyversion option for doctest (issue 3303)

tk0miya commented on sphinx/ext/doctest.py:
```diff
@@ -23,6 +24,8 @@
 from docutils import nodes
 from docutils.parsers.rst import Directive, directives
 
+from distutils.version import StrictVersion as V
```
Why abbrviate? `V` is not clear name to me.

Anyway, it should be imported after importing six.
We usually import libraries in order of python standards, 3rd parties, docutils and sphinx's.

tk0miya commented on sphinx/ext/doctest.py:
```diff
@@ -103,10 +106,40 @@ def run(self):
                 flag = doctest.OPTIONFLAGS_BY_NAME[option[1:]]  # type: ignore
                 node['options'][flag] = (option[0] == '+')
+        if self.name == 'doctest' and 'pyversion' in self.options:
+            try:
```
Could you separate this into a single function and add some test?
It seems the evaluation code are not tested well.

tk0miya reviewed (commented):
LGTM with nits

tk0miya commented:
Ah, I'd overlooked about travis.
It seems CI process was failed.
Please check the logs of travis ci and fix them.

marco-buttu committed:
Added pyversion option for doctest (issue 3303)

marco-buttu commented:
Thanks for your time @tk0miya . I am not completely satisfied about the test, because I do not know how to propery create a TestDirective object, so I gave it dummy arguments...

tk0miya commented on sphinx/ext/doctest.py:
```diff
@@ -107,41 +106,52 @@ def run(self):
+        """
+        operands = ('<=', '<', '==', '>=', '>')
+        if operand not in operands:
+            self.document.reporter.warning(
```
Is this typo for self. **state** .reporter?
Did you check this warning works surely?

marco-buttu replied:
Your interface is just perfect :-)  So, I will change my interface following your suggestion, fix the typo and check (with pytest) everything works properly. Thank you very much!

marco-buttu committed:
Added pyversion option for doctest (issue 3303)

Code, tests and documentation.

tk0miya reviewed (commented):
If you are not satisfied the tests, how about changing the interface like following?
```
def compare_version(ver1, ver2, operand):
    if operand not in operands:
        raise ValueError  # raise ValueError if unknown operand

    ...
    return result  # return boolean value as a result of comparison


class TestDirective(Directive):
    def run(self):
        ...
        if self.name == 'doctest' and 'pyversion' in self.options:
            try:
                if compare_version(platform.python_version(), version, operand):
                    ...
            except ValueError:
                # handle error for unknown operand and other reasons.
                self.state.document.reporter.warning(...)
```

marco-buttu committed:
compare_version() replaces proper_pyversion()

marco-buttu committed:
Removed useless import

marco-buttu committed:
Comparison made by distutils.version.LooseVersion

tk0miya commented on doc/ext/doctest.rst:
```diff
@@ -73,6 +73,19 @@ a comma-separated list of group names.
+
+         .. doctest::
+            :pyversion: > 3.3
+ 
```
style-check script says
```
doc/ext/doctest.rst:82: trailing whitespace
```

Please check Travis CI's status after fixed up.

marco-buttu replied:
I am sorry @tk0miya , I checked Travis but I was not able to figure out where the problem was, and I thought to fix it on this morning. I have not enough confidence with all the tools yet :-/

tk0miya reviewed (approved):
LGTM with nits

marco-buttu committed:
Fixed flake8 style check

marco-buttu commented:
I double checked everything, the warning messages too, and it works properly. Thank you very very much @tk0miya

tk0miya commented:
LGTM!
I will merge this if no objection in a few days.

marco-buttu committed:
Merge remote-tracking branch 'upstream/master'

marco-buttu committed:
Added author name and changes (Sphinx 1.5.3)

marco-buttu committed:
Preserved alphabetically order in author list

marco-buttu committed:
In case of wrong option, warn the right reason.

For instance, in case of `:options: +SKIPO` the message
was "missing '+' or '-' in +SKIPO option`. Now it will be
"SKIPO is not a valid option."

tk0miya commented on CHANGES:
```diff
@@ -73,6 +73,8 @@ Deprecated
 Features added
 --------------
 
+* #3303: Added ``:pyversion:`` option to the doctest directive
```
Sorry, this feature will be released as 1.6.
1.5.x series is under maintenance period. so only bug fixes are allowed.

tk0miya commented on doc/ext/doctest.rst:
```diff
@@ -73,6 +73,19 @@ a comma-separated list of group names.
+     comparison is performed by `distutils.version.LooseVersion
+     <https://www.python.org/dev/peps/pep-0386/#distutils>`__.
+
+     .. versionadded:: 1.5
```
Please rewrite this to 1.6.

marco-buttu replied:
:+1:

tk0miya commented on sphinx/ext/doctest.py:
```diff
@@ -101,12 +126,32 @@ def run(self):
+                on_or_off, option_name = option[0], option[1:]
+                if on_or_off not in '+-':  # type: ignore
+                    self.state.document.reporter.warning(
+                        "missing '+' or '-' in '%s' option." % option,
```
Could you wrap with `_(...)` function?
Ex:
```
from sphinx.locale import _
self.state.document.reporter.warning(_("missing '+' or '-' in '%s' option.") % option)
```

Then it makes the message translatable.

marco-buttu replied:
:+1:

tk0miya commented on sphinx/ext/doctest.py:
```diff
@@ -101,12 +126,32 @@ def run(self):
-                if (option[0] not in '+-' or option[1:] not in
-                        doctest.OPTIONFLAGS_BY_NAME):  # type: ignore
-                    # XXX warn?
+                on_or_off, option_name = option[0], option[1:]
```
I feel `on_or_off` is a bit ugly name. How about `prefix`, `sign` or others?

marco-buttu replied:
``on_or_off`` is the name given to ``+/-`` in the doctest documentation, but I agree with you that is not meaningful. I wil use ``prefix``, thanks for the suggestion :-)

tk0miya replied:
I don't know that. But it looks better to me :-)

tk0miya reviewed (changes requested):
I reviewed again for latest update.
Could you check them please?

marco-buttu committed:
From version 1.5 to 1.6. Warnings wrapped with locale.

marco-buttu commented:
Done :) Thanks @tk0miya

tk0miya committed:
Merge branch 'master' into master

tk0miya commented:
Merged!
Thank you for contribution :-)
